src/uri.cr

```diff
@@ -128,7 +128,7 @@ class URI
   def full_path
     String.build do |str|
       str << (@path.try { |p| !p.empty? } ? @path : "/")
-      str << "?" << @query if @query
+      str << "?" << @query if (query = @query) && !query.empty?
```

Comment:
I would personally prefer move the assign and condition to the left side (a regular `if`)

```crystal
if (query = @query) && !query.empty?
  str << "?" << query
end
```

Reply:
You should probably replace this with the following which is much clearer as to your intent.
```crystal
str << "?" << @query if !@query.to_s.empty?
```
**More importantly** since `@query` is literally the same thing as `query` owing to the `getter` setting it to itself is just weird. I wanted to see if there was some dark magic I was ignorant of, but as far as i can tell this assignment makes no sense.

```
class Foo
  getter x
  setter x
  def initialize(@x : String?)
  end
  def test : Bool
    (x = @x) ? true : false
  end
  def test2 : Bool
    x ? true : false
  end
  def test3 : Bool
     @x ? true : false
  end

end

```
then checking it in icr...

```
icr(0.23.1) > f = Foo.new("something")
 => #<Foo:0x107231e60 @x="something">
icr(0.23.1) > f.test
 => true
icr(0.23.1) > f.test2
 => true
icr(0.23.1) > f.test3
 => true
icr(0.23.1) > f = Foo.new(nil)
 => #<Foo:0x1070c8e40 @x=nil>
icr(0.23.1) > f.test
 => false
icr(0.23.1) > f.test2
 => false
icr(0.23.1) > f.test3
 => false

```

Reply:
@masukomi Thanks for the feedback and the detailed response. That example is not *quite* right for showing the problem. Here's one that illustrates the issue: https://play.crystal-lang.org/#/r/36hx

```crystal
class Foo
  @query : String?
  
  def full_path
    String.build do |str|
      str << @query if @query && @query.empty?
    end
  end
end

Foo.new.full_path
```

Results in:

```crystal
in line 6: undefined method 'empty?' for Nil (compile-time type is (String | Nil))
```

I wish I could find the documentation for this, but the gist is that instance variable can change at any time so you need to set it to a local variable *first* so the compiler can be sure that it is *still* not nil. Does that make sense?

`!@query.to_s.empty?` would work, but it requires making an intermediate string object if the query is `nil`, which IMO is not worth it for slighly improved readability